src/fit.jl

```diff
@@ -147,76 +153,72 @@ function eval_metrics!(
     estimator::LGBMEstimator,
     tests_names::Vector{String},
     iter::Integer,
-    metrics_store_idx::Integer,
-    metrics_stored_count::Integer,
-    n_metrics::Integer,
     verbosity::Integer,
-    bigger_is_better::Vector{Float64},
-    best_score::Matrix{Float64},
-    best_iter::Matrix{Int},
+    bigger_is_better::Dict{String,Float64},
+    best_score::Dict{String,Dict{String,Real}},
+    best_iter::Dict{String,Dict{String,Real}},
     metrics::Vector{String},
 )
+    now_scores = Dict{String,Vector{Float64}}()
 
     if (iter - 1) % estimator.metric_freq == 0
         if estimator.is_training_metric
-            scores = LGBM_BoosterGetEval(estimator.booster, 0)
-            store_scores!(results, estimator, metrics_store_idx, metrics_stored_count, "training", scores, metrics)
-            print_scores(estimator, iter, "training", n_metrics, scores, metrics, verbosity)
+            now_scores["training"] = LGBM_BoosterGetEval(estimator.booster, 0)
+        end
+        for (test_idx, tests_name) in enumerate(tests_names)
+            now_scores[tests_name] = LGBM_BoosterGetEval(estimator.booster, test_idx)
         end
     end
 
-    # Metrics for test sets
-    if (iter - 1) % estimator.metric_freq == 0 || estimator.early_stopping_round > 0
-        for (test_idx, test_name) in enumerate(tests_names)
-            scores = LGBM_BoosterGetEval(estimator.booster, test_idx)
-            # Check if progress should be stored and/or printed
-            if (iter - 1) % estimator.metric_freq == 0
-                store_scores!(results, estimator, metrics_store_idx, metrics_stored_count, test_name, scores, metrics)
-                print_scores(estimator, iter, test_name, n_metrics, scores, metrics, verbosity)
-            end
-
-            # Check if early stopping is called for
-            @inbounds for metric_idx in eachindex(metrics)
-                maximize_score = bigger_is_better[metric_idx] * scores[metric_idx]
-                if maximize_score > best_score[metric_idx, test_idx]
-                    best_score[metric_idx, test_idx] = maximize_score
-                    best_iter[metric_idx, test_idx] = iter
-                elseif iter - best_iter[metric_idx, test_idx] >= estimator.early_stopping_round
-                    # This will shrink it up to the current stored metric
-                    shrinkresults!(results, metrics_store_idx)
-                    log_info(verbosity, "Early stopping at iteration ", iter,
-                             ", the best iteration round is ", best_iter[metric_idx, test_idx], "\n")
-                    return 1
-                end
+    # check early stopping condition
+    if estimator.early_stopping_round > 0
```

Comment:
Is it right that metrics are grabbed at every iteration if early stopping is on at all?

Reply:
that would depend on the criteria `(iter - 1) % estimator.metric_freq == 0` in line 161, right?  If that condition is not met, now_scores will be an empty dict, and nothing will really be stored/printed